src/history/hs_rec.c

```diff
@@ -213,10 +217,21 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
         error_on_ts_ordering = false;
     }
 
-    if (ret == 0)
-        WT_ERR(__hs_delete_reinsert_from_pos(session, cursor, btree->id, key, tw->start_ts + 1,
-          true, false, error_on_ts_ordering, &counter, tw));
+    if (ret == 0) {
+        WT_ASSERT(session, tw->start_ts + 1 > WT_TS_NONE);
+        /* Get the history store cursor position. */
+        WT_ERR_NOTFOUND_OK(__hs_cursor_pos(session, cursor, btree->id, key, tw->start_ts + 1,
+                             error_on_ts_ordering, tw, &non_ts_updates),
+          true);
 
+        /*
+         * Skip reinsert for non-timestamped updates. Only proceed when a valid cursor position is
+         * found, otherwise there is nothing to do, skip this function.
+         */
+        if (!non_ts_updates && ret == 0)
+            WT_ERR(__hs_delete_reinsert_from_pos(session, cursor, btree->id, key, tw->start_ts + 1,
```

Comment:
We should explain why we don't need to delete and reinsert when we are only in non-timestamp land.

Reply:
This change has been factored out and discussed offline.